Remove LRU cache from ResourceQuota admission plugin

AwesomePatrol opened the pull request:
/kind bug
/kind regression

#### What this PR does / why we need it:

This PR mitigates the problem of high latency caused by consistent reads made in Admission/Validation of ResourceQuota. LISTs made with Informer provide sufficient consistency guarantees. More details are available in the linked bug report.

#### Which issue(s) this PR fixes:

Issue #129931 

#### Special notes for your reviewer:

Tests in `resource_access_test` focused on caching and avoiding duplicated LIST requests. As I deleted the code, the tests became obsolete. `GetQuotas` is still covered by other tests in the module.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/cc @jpbetz @serathius @liggitt @deads2k

liggitt commented:
/ok-to-test

Update the release note to indicate this resolves a performance regression of write requests related to the ConsistentListFromCache feature gate, enabled in 1.31+ by default.

liggitt commented:
/lgtm
/approve

/hold for ack from second api-machinery reviewer (@deads2k?)

deads2k commented:
> /hold for ack from second api-machinery reviewer (@deads2k?)

lgtm

/hold cancel

seans3 commented:
/triage accepted

liggitt commented:
/hold

presubmit unit test failures look relevant, those might need adjusting to ensure the lister contains the resource quota under test

presubmit verify failure is legitimate, drop the unused type:
```
staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:82:6: type `liveLookupEntry` is unused (unused)
type liveLookupEntry struct {
```

presubmit integration failures do not look relevant... @jpbetz can you make sure there's an issue for the `k8s.io/kubernetes/test/integration/apiserver/cel` failures?

jpbetz commented:
> presubmit integration failures do not look relevant... @jpbetz can you make sure there's an issue for the `k8s.io/kubernetes/test/integration/apiserver/cel` failures?

Got a match: https://github.com/kubernetes/kubernetes/issues/125555. I'll follow up.

liggitt commented on plugin/pkg/admission/resourcequota/admission_test.go:
```diff
@@ -1251,14 +1253,13 @@ func TestAdmitLimitedResourceWithMultipleQuota(t *testing.T) {
+	kubeClient := fake.NewClientset(resourceQuota1, resourceQuota2)
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, 0)
+	informerFactory.Core().V1().ResourceQuotas().Informer().GetStore().Add(resourceQuota1)
+	informerFactory.Core().V1().ResourceQuotas().Informer().GetStore().Add(resourceQuota2)
```
No, just fix the lint issues with code you modified

liggitt commented on plugin/pkg/admission/resourcequota/admission_test.go:
```diff
@@ -1297,13 +1299,14 @@ func TestAdmitLimitedResourceWithQuotaThatDoesNotCover(t *testing.T) {
 	defer close(stopCh)
 
 	informerFactory := informers.NewSharedInformerFactory(kubeClient, 0)
+	if err := informerFactory.Core().V1().ResourceQuotas().Informer().GetStore().Add(resourceQuota); err != nil {
+		t.Fatalf("Failed to add ResourceQuota(%q) to informer: %v", resourceQuota.ObjectMeta.Name, err)
+	}
```
rather than directly manipulating the store, I think we should do the following after calling `createHandlerWithConfig`, in the same place we are calling `indexer.Add` today:

```
       informerFactory.Start(stopCh)
       informerFactory.WaitForCacheSync(stopCh)
```

liggitt replied:
same comment applies to all places

AwesomePatrol replied:
Started the informer in `createHandlerWithConfig` that let me remove all references to Indexer and Store. Now it is much easier to keep track of `ResourceQuota`s under test, because they need to be passed to `fake.Clientset` only

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/resource_access.go:
```diff
@@ -112,34 +101,6 @@ func (e *quotaAccessor) GetQuotas(namespace string) ([]corev1.ResourceQuota, err
 		return nil, fmt.Errorf("error resolving quota: %v", err)
 	}
 
-	// if there are no items held in our indexer, check our live-lookup LRU, if that misses, do the live lookup to prime it.
-	if len(items) == 0 {
-		lruItemObj, ok := e.liveLookupCache.Get(namespace)
-		if !ok || lruItemObj.(liveLookupEntry).expiry.Before(time.Now()) {
```
I thought of a case where this removal *might* make a noticeable difference in behavior: on a cold start of an API server, where `lister` has not completed its initial sync, but there are pre-existing resourcequota objects a live lookup would find.

I wonder if keeping the live fallback but narrowing it to the case where the lister returns no items _and_ hasn't completed a sync yet would be a better idea:

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go b/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go
index e7d59bb707d..5455b414eda 100644
--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go
@@ -114,7 +114,9 @@ func (a *QuotaAdmission) SetExternalKubeClientSet(client kubernetes.Interface) {
 
 // SetExternalKubeInformerFactory registers an informer factory into QuotaAdmission
 func (a *QuotaAdmission) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
-	a.quotaAccessor.lister = f.Core().V1().ResourceQuotas().Lister()
+	quotas := f.Core().V1().ResourceQuotas()
+	a.quotaAccessor.lister = quotas.Lister()
+	a.quotaAccessor.hasSynced = quotas.Informer().HasSynced
 }
 
 // SetQuotaConfiguration assigns and initializes configuration and evaluator for QuotaAdmission
@@ -144,6 +146,9 @@ func (a *QuotaAdmission) ValidateInitialization() error {
 	if a.quotaAccessor.lister == nil {
 		return fmt.Errorf("missing quotaAccessor.lister")
 	}
+	if a.quotaAccessor.hasSynced == nil {
+		return fmt.Errorf("missing quotaAccessor.hasSynced")
+	}
 	if a.quotaConfiguration == nil {
 		return fmt.Errorf("missing quotaConfiguration")
 	}
diff --git a/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/resource_access.go b/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/resource_access.go
index d189446f032..fd4c102e6d5 100644
--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/resource_access.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/resource_access.go
@@ -48,6 +48,9 @@ type quotaAccessor struct {
 	// lister can list/get quota objects from a shared informer's cache
 	lister corev1listers.ResourceQuotaLister
 
+	// hasSynced indicates whether the lister has completed its initial sync
+	hasSynced func() bool
+
 	// liveLookups holds the last few live lookups we've done to help ammortize cost on repeated lookup failures.
 	// This lets us handle the case of latent caches, by looking up actual results for a namespace on cache miss/no results.
 	// We track the lookup result here so that for repeated requests, we don't look it up very often.
@@ -112,8 +115,8 @@ func (e *quotaAccessor) GetQuotas(namespace string) ([]corev1.ResourceQuota, err
 		return nil, fmt.Errorf("error resolving quota: %v", err)
 	}
 
-	// if there are no items held in our indexer, check our live-lookup LRU, if that misses, do the live lookup to prime it.
-	if len(items) == 0 {
+	// if there are no items held in our unsynced lister, check our live-lookup LRU, if that misses, do the live lookup to prime it.
+	if len(items) == 0 && !e.hasSynced() {
 		lruItemObj, ok := e.liveLookupCache.Get(namespace)
 		if !ok || lruItemObj.(liveLookupEntry).expiry.Before(time.Now()) {
 			// use singleflight.Group to avoid flooding the apiserver with repeated
```

That would limit the impact to writes sent to a still-initializing server... once it's resourcequota informer was synced, live lookup would never be used again.

liggitt replied:
Long term, after merging / backporting ^, I think it could make sense to require the resourcequota lister to have synced before the resourcequota admission plugin allows any namespaced writes, but we can work through that in a follow-up

liggitt replied:
SetReadyFunc() is probably what I'd recommend in 1.33 / master, yes. You would also have to call WaitForReady() at the right place in Validate() (it doesn't look like limitranger is doing that, so calling SetReadyFunc() isn't accomplishing anything)

Before doing that, please restore the live lookup and limit it to when the informer has not yet synced. We will backport that change to 1.31 and 1.32.

AwesomePatrol replied:
I need to first verify if this code matters: https://github.com/kubernetes/kubernetes/blob/e9c9be4007d1664e68796af02b8978640d2c1b26/staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/plugin.go#L194-L196

AwesomePatrol replied:
> You would also have to call WaitForReady() at the right place in Validate() (it doesn't look like limitranger is doing that, so calling SetReadyFunc() isn't accomplishing anything)

This seems like a very common mistake. In this PR I will fix ResourceQuota/LimitRanger, but there are still many other left:
* storageclass
* persistentvolume
* serviceaccount
* podsecurity
* runtimeclass
* priority
* podtolerationrestriction
* defaultingressclass

Do you want me to fix it in this PR, a different one, refactor in a new PR, or leave it as is?

liggitt commented:
/approve cancel

AwesomePatrol committed:
Remove additional LIST and LRU cache from ResourceQuota

The informer keeps track of the data already so there is no need for an
additional layer of caching and LISTing. We add wait to ResourceQuota
admission hook to make sure it is synchronized before processing
requests.

AwesomePatrol committed:
Remove references to Indexer and Store in admission test

ResourceQuota's Admission tests now start an informer that fetches
resources from the fake kubeClient.

liggitt commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -163,50 +152,19 @@ func (l *LimitRanger) GetLimitRanges(a admission.Attributes) ([]*corev1.LimitRan
 	}
 
-	// if there are no items held in our indexer, check our live-lookup LRU, if that misses, do the live lookup to prime it.
-	if len(items) == 0 {
```
scope this PR to just the resourcequota changes... dropping the live lookup on limitranger has actual user-visible functional implications during rapid initialization I'm not as sure about, let's move that to a distinct PR to think through

AwesomePatrol replied:
Done.

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:
```diff
@@ -164,6 +154,10 @@ func (a *QuotaAdmission) Validate(ctx context.Context, attr admission.Attributes
 		return nil
 	}
+	// we need to wait for our caches to warm
+	if !a.WaitForReady() {
```
doing this here, before calling Evaluate(), means that in unsynced cases, we'll reject resources we would otherwise have ignored:

https://github.com/kubernetes/kubernetes/blob/d36737c817303e525e00e5ada24a7a79c05ebb53/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go#L601-L606

or skipped:

https://github.com/kubernetes/kubernetes/blob/d36737c817303e525e00e5ada24a7a79c05ebb53/staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go#L617-L621

I think the WaitForReady call should happen *after* both of those, so we only block on sync for things quota is actually interested in. That will mean plumbing the wait function down into the evaluator to invoke.

AwesomePatrol replied:
I moved it to the evaluator, but I am not sure if this is the best place. `init` can't be used as it doesn't let for retrying errors. Moving to `quotaAccessor` may work, but it would likely need a custom implementation of `waitForReady`:
```go
func (a *quotaAccessor) waitForReady() error {
  if a.hasSynced() {
    return nil
  }
  if err := wait.PollUntilContext(context.Background, 100 * time.Millisecond, time.Second, func(_ context.Context) { return a.hasSynced(), nil }); err != nil {
    return fmt.Errorf("not ready")
  }
```

I also hope that compiler manages to optimize handler's `WaitForReady` as otherwise it creates a timer only to be GCed right after (in cases when `readyFunc` returns always `true`, i.e. post sync)

AwesomePatrol commented:
Now that v1.34 has been released. Can we try to clean it up?

/reopen
/remove-lifecycle rotten

AwesomePatrol commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129998/pull-kubernetes-e2e-gce/1964952549702242304 is a real test failure I need to address first
